working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+
+The changes we need are pretty minimal and could be quick to perform manually, but for the practice, let's see how Copilot handles the change. If we highlight the `validate_book` function, open up the inline Copilot chat, and use the prompt:
+
+> Please update this function so that we could use it for any model class
+
+The response delivers exactly what we're looking for:
+```
+def validate_model(model_class, model_id):
+    try:
+        model_id = int(model_id)
+    except:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} invalid"}, 400))
+
+    model = model_class.query.get(model_id)
+
+    if not model:
+        abort(make_response({"message": f"{model_class.__name__.lower()} {model_id} not found"}, 404))
+
+    return model
+```
+
+Our function is updated, now we just need to update our dependent functions and tests to use `validate_model` instead of `validate_book`, which includes passing in the new `model_class` parameter. We can do this manually using VS Code's Find & Replace tools then adding in our new parameter, or we can reach out to Copilot. 
+
+One option is to open each file, select the whole file, and ask Copilot to change any instances of `validate_book` to `validate_model`. We need to carefully review the lines Copilot suggests changing to ensure nothing unexpected is altered, but it can save us a little time since we need to do slightly more than find and replace the function name.
+
+In our experimentation, we found that in the `routes.py` file, we could use a prompt with very few details to get the changes that we wanted: 
+
+> Please replace all uses of validate_book with validate_model
+
+However, in the `test_routes.py` file, we had to add more detail for Copilot to recognize that there was a new class parameter which was required wherever the function was invoked, and that we needed import the `Book` class:
+
+> Please replace all uses of validate_book(book_id) with validate_model(model_class, model_id). Import any required model classes.
+
+## Summary
+
+This hands-on experience in a project wraps up our whirlwind tour of GitHub's Copilot! Whether we're writing new code or working in an existing base Copilot can help us with tasks like generating ideas for tests, surfacing error handling we may have missed, and generally writing code quicker. Even if Copilot's suggestions look good at a glance, we always need to carefully proofread them both for correctness and completeness. Just because a code suggestion works does not mean it fits all of our needs, so we need to remain vigilant and critical of what Copilot provides. If you're interested in more practice with Copilot, we highly suggest revisiting Solar System, Task List, Inspo Board, or any other project that you're already familiar with so you know what the end product could look like and can focus on working with Copilot.
```

Comment:
- split the summary from the more practice
- base → codebase
- requires comparative adverb
- if → when
- spell out Inspiration
```suggestion
This hands-on experience in a project wraps up our whirlwind tour of GitHub's Copilot! Whether we're writing new code or working in an existing codebase, Copilot can help us with tasks like generating ideas for tests, surfacing error handling we may have missed, and generally writing code more quickly. Even when Copilot's suggestions look good at a glance, we always need to carefully proofread them both for correctness and completeness. Just because a code suggestion works does not mean it fits all of our needs, so we need to remain vigilant and critical of what Copilot provides.

If you're interested in more practice with Copilot, we highly suggest revisiting Solar System, Task List, Inspiration Board, or any other project that you're already familiar with so you know what the end product could look like and can focus on working with Copilot.
```